services/QuillLMS/app/controllers/sessions_controller.rb

```diff
@@ -104,4 +82,43 @@ def finish_sign_up
     end
     redirect_to profile_path
   end
+
+  def find_login_error
+    error_conditions = {
+      -> { @user.nil? || @user.sales_contact? } =>
+        { message: 'An account with this email or username does not exist. Try again.', type: 'email' },
+      -> { @user.google_id && @user.password_digest.nil? } =>
+        { message: 'Your account is linked to Google. To sign in with your email address, create a password by ' \
+                   "clicking 'Forgot password'.", type: 'email' },
+      -> { @user.clever_id } =>
+        { message: 'Oops! You have a Clever account. Log in that way instead.', type: 'email' },
+      -> { @user.classlink_id } =>
+        { message: 'Oops! You have a ClassLink account. Log in that way instead.', type: 'email' },
+      -> { @user.password_digest.nil? } =>
+        { message: 'Something went wrong verifying your password. Please use the "Forgot password?" link below ' \
+                   'to reset it.', type: 'email' }
+    }
+
+    error_conditions.find { |condition, _response| condition.call }&.last
```

Comment:
I had not realized that `{}.find` would return a 2-array of `[key, value]`.  Today I learned!